openedx/core/djangoapps/programs/tasks.py

```diff
@@ -24,32 +124,35 @@ def award_program_certificates(username):
     This task may also be invoked independently of any course completion status
     change - for example, to backpopulate missing program credentials for a
     user.
-
-    TODO: this is shelled out and incomplete for now.
     """
 
+    student = User.objects.get(username=username)
+
     # fetch the set of all course runs for which the user has earned a certificate
-    LOGGER.debug('fetching all completed courses for user %s', username)
-    user_certs = get_certificates_for_user(username)
-    course_certs = [
-        {'course_id': uc['course_id'], 'mode': uc['mode']}
-        for uc in user_certs
-        if uc['status'] in ('downloadable', 'generating')
-    ]
+    course_certs = get_completed_courses(student)
 
     # invoke the Programs API completion check endpoint to identify any programs
     # that are satisfied by these course completions
-    LOGGER.debug('determining completed programs for courses: %r', course_certs)
-    program_ids = []  # TODO
+    programs_client = get_api_client(ProgramsApiConfig.current(), student)
+    program_ids = get_completed_programs(programs_client, course_certs)
 
     # determine which program certificates the user has already been awarded, if
     # any, and remove those, since they already exist.
-    LOGGER.debug('fetching existing program certificates for %s', username)
-    existing_program_ids = []  # TODO
+    existing_program_ids = get_awarded_certificate_programs(student)
+
     new_program_ids = list(set(program_ids) - set(existing_program_ids))
 
     # generate a new certificate for each of the remaining programs.
-    LOGGER.debug('generating new program certificates for %s in programs: %r', username, new_program_ids)
-    for program_id in new_program_ids:
-        LOGGER.debug('calling credentials service to issue certificate for user %s in program %s', username, program_id)
-        # TODO
+    if new_program_ids:
+        LOGGER.debug('generating new program certificates for %s in programs: %r', username, new_program_ids)
+        credentials_client = get_api_client(
+            CredentialsApiConfig.current(),
+            User.objects.get(username='FIXME-NO-SERVICE-USER')
+        )
+        for program_id in new_program_ids:
+            LOGGER.debug(
+                'calling credentials service to issue certificate for user %s in program %s',
+                username,
+                program_id,
+            )
+            award_program_certificate(credentials_client, student, program_id)
```

Comment:
i was doing that to be consistent with other helper methods in this module which need the User object, but i suppose the fewer places we do that, the better.  will change.